### PR TITLE
feat: switch from lowercase-keys to caseless for headers

### PR DESCRIPTION
**Pull request by forivall**
BREAKING CHANGE: Plugins, hooks, etc. could break with this change, as NormalizedOptions.headers will no longer always be lowercase.
Among other possible breaking issues

I needed this to correctly call an api where, unfortunately, the people who implemented it check the headers in a case sensitive manner. 😢 Yes, i know RFC 2616 specifies that they're case insensitive. Yes, i tried to tell the people who implemented the api. Alas, this is my curse.

So, i forked `got`, shoved in `caseless` where `lowercase-keys` once was, and it works! This could definitely be done in a different way, but it works, and all the tests pass, and i can call that API. Published at [got-caseless](https://npm.im/got-caseless).

I'm open to discussion, further changes, swapping out caseless for something cleaner, smaller, more lean, etc.

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [ ] I have included some tests.
  - There's one test I added, and it works in my code. Definitely could use more.
- [ ] If it's a new feature, I have included documentation updates.
  - Is this a new feature? 🤷‍♀  In any case, it's a breaking change, and should only be merged on a major version bump.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -5,7 +5,8 @@
 		"target": "es2018", // Node.js 10
 		"lib": [
 			"es2018"
-		]
+		],
+		"sourceMap": true
```
Comment (forivall) [inline tsconfig.json]: Oops. This should/could be removed - I added it so i could jump to the failing tests a wee bit quicker. I could use `tsc --sourceMap` in the future.

**Comment by forivall** [comment]
... and as I noticed the rework branch, (#1051), I've prepared a branch on top of that instead: https://github.com/szmarczak/got/compare/rework...forivall:rework-caseless

**Comment by szmarczak** [comment]
I appreciate the effort, but there's no need for this change. You can modify the response using hooks.

**Comment by forivall** [comment]
I tried using hooks. Got went back and lowercased the request headers. I'll submit a test pull request.

**Comment by forivall** [comment]
What would work (and I only thought of this now) is to supply a custom request function that recases the headers (my first attempt was a beforeRequest hook, but they get re-lowercased)

**Comment by forivall** [comment]
Most of this is typing that appeases typescript, but this is the start of it.
```ts
function gotFixHeaderCasing(canonicalHeaders: string[]): ExtendOptions {
  const isUrlInstance = (require('@sindresorhus/is') as typeof import('@sindresorhus/is'))
    .default.urlInstance;
  const extendOpts: ExtendOptions = {};
  const hook: BeforeRequestHook = (options) => {
    const inner = options[requestSymbol];
    type RequestOptions = import('http').RequestOptions;
    type IncomingMessage = import('http').IncomingMessage;
    type ClientRequest = import('http').ClientRequest;
    function request(
      options: RequestOptions | string | URL,
      callback?: (res: IncomingMessage) => void
    ): ClientRequest;
    function request(
      url: string | URL,
      options: RequestOptions,
      callback?: (res: IncomingMessage) => void
    ): ClientRequest;
    function request(
      ...args: Parameters<
        (
          urlOrOptions: RequestOptions | string | URL,
          optionsOrCallback?: RequestOptions | ((res: IncomingMessage) => void),
          maybeCallback?: (res: IncomingMessage) => void
        ) => void
      >
    ) {
      const [urlOrOptions, optionsOrCallback] = args;
      const requestOptions =
        optionsOrCallback && typeof optionsOrCallback !== 'function'
          ? optionsOrCallback
          : typeof urlOrOptions !== 'string' && !isUrlInstance(urlOrOptions)
          ? urlOrOptions
          : null;
      if (requestOptions) {
        // TODO: re-case headers
      }
      return inner(...(args as Parameters<typeof inner>));
    }
    options.request = request;
  };
  extendOpts.hooks = {
    beforeRequest: [hook],
  };
  return extendOpts;
}
```

**Comment by forivall** [comment]
newer & complete version for got@11:

```ts
/*
(c) 2020 Emily Marigold Klassen <forivall@gmail.com>
ISC License
*/
import type { ExtendOptions, InitHook, Headers } from 'got';

export const defaultCanonicalHeaderCasing: ReadonlyArray<string> = [
  'User-Agent',
  'Content-Length',
  'Accept-Encoding',
];

const makeLowercaseMapping = (headers: ReadonlyArray<string>) =>
  headers.reduce((mapping, h) => {
    mapping[h.toLowerCase()] = h;
    return mapping;
  }, {} as Record<string, string>);

export default function gotFixHeaderCasing(
  canonicalHeaders: ReadonlyArray<string> = defaultCanonicalHeaderCasing
): ExtendOptions {
  const canonicalHeaderMapping = makeLowercaseMapping(canonicalHeaders);
  const extendOpts: ExtendOptions = {};
  const initHook: InitHook = (rawOptions) => {
    const origHeaderMapping = {
      ...canonicalHeaderMapping,
      ...makeLowercaseMapping(Object.keys(rawOptions.headers || {})),
    };
    rawOptions.request = (dummyUrl, opts): undefined => {
      opts.headers = Object.keys(opts.headers as Headers).reduce(
        (remapped, h) => {
          remapped[origHeaderMapping[h] || h] = (opts.headers as Headers)[h];
          return remapped;
        },
        {} as Headers
      );
      return;
    };
  };
  extendOpts.hooks = {
    init: [initHook],
  };
  return extendOpts;
}
```

I'll publish this on the weekend.

**Comment by Yopi** [comment]
I had to modify the hook slightly to get it to work:

I'm not entirely sure why rawOptions is sometimes undefined, but it seems like there might be an issue with making requests when not specifying any options to the request (i.e. just `got(url);`)

```
/*
(c) 2020 Emily Marigold Klassen <forivall@gmail.com>
ISC License
*/
import type {
  InitHook, Headers, Hooks, Options,
} from 'got';

export const defaultCanonicalHeaderCasing: ReadonlyArray<string> = [
  'User-Agent',
  'Content-Length',
  'Accept-Encoding',
];

const makeLowercaseMapping = (headers: ReadonlyArray<string>): Record<string, string> =>
  headers.reduce((mapping, h) => {
    mapping[h.toLowerCase()] = h;
    return mapping;
  }, {} as Record<string, string>);

export default function gotFixHeaderCasing(
  canonicalHeaders: ReadonlyArray<string> = defaultCanonicalHeaderCasing,
): InitHook {
  const canonicalHeaderMapping = makeLowercaseMapping(canonicalHeaders);
  return (rawOptions: Options): void => {
    if (rawOptions == null) {
      rawOptions = {};
    }

    const origHeaderMapping = {
      ...canonicalHeaderMapping,
      ...makeLowercaseMapping(Object.keys(rawOptions.headers || {})),
    };
    // eslint-disable-next-line no-multi-assign
    const hooks: Hooks = (rawOptions.hooks = { ...rawOptions.hooks });
    hooks.beforeRequest = (hooks.beforeRequest ?? []).concat((opts) => {
      opts.headers = Object.keys(opts.headers as Headers).reduce(
        (remapped, h) => {
          remapped[origHeaderMapping[h] || h] = (opts.headers as Headers)[h];
          return remapped;
        },
        {} as Headers,
      );
    });
  };
}
```

**Pull request closed without merging**